compiler/rustc_middle/src/ty/relate.rs

```diff
@@ -641,6 +641,36 @@ pub fn super_relate_consts<'tcx, R: TypeRelation<'tcx>>(
                 ty: a.ty(),
             }));
         }
+        (ty::ConstKind::Expr(ae), ty::ConstKind::Expr(be)) => {
+            let r = relation;
+            // FIXME(julianknodt): is it possible to relate two consts which are not identical
+            // exprs? Should we care about that?
+            let expr = match (ae, be) {
+                (Expr::Binop(a_op, al, ar), Expr::Binop(b_op, bl, br)) if a_op == b_op => {
+                    Expr::Binop(a_op, r.consts(al, bl)?, r.consts(ar, br)?)
+                }
+                (Expr::UnOp(a_op, av), Expr::UnOp(b_op, bv)) if a_op == b_op => {
+                    Expr::UnOp(a_op, r.consts(av, bv)?)
+                }
+                (Expr::Cast(ak, av, at), Expr::Cast(bk, bv, bt)) if ak == bk => {
+                    Expr::Cast(ak, r.consts(av, bv)?, r.tys(at, bt)?)
+                }
+                (Expr::FunctionCall(af, aa), Expr::FunctionCall(bf, ba))
+                    if aa.len() == ba.len() =>
+                {
+                    let func = r.consts(af, bf)?;
+                    let mut related_args = Vec::with_capacity(aa.len());
+                    for (a_arg, b_arg) in aa.iter().zip(ba.iter()) {
+                        related_args.push(r.consts(a_arg, b_arg)?);
+                    }
+                    let related_args = List::from_arena(tcx.arena, &related_args);
```

Comment:
```suggestion
                    let related_args = tcx.intern_const_list(&related_args);
```
(or however the method on `TyCtxt` is called)

`List` from arena is pretty much always incorrect outside of the interners themselves.

`List` assumes uniqueness of its content so we have to always go through the interner to guarantee that this stays that way